File: app/src/main/java/com/instacart/android/challenges/MainActivity.java
package com.instacart.android.challenges;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class MainActivity extends AppCompatActivity {

    private Toolbar toolbar;
    private RecyclerView recyclerView;
    private ItemAdapter adapter;

    @Override protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        View itemScreenContainerView = findViewById(R.id.item_screen_container);
        bindViews(itemScreenContainerView);

        MainActivityViewModel viewModel = new MainActivityViewModel();
        viewModel.setStateUpdateListener(this::renderItemList);
    }

    private void renderItemList(ItemListViewState state) { }

    private void bindViews(View parent) {
        toolbar = parent.findViewById(R.id.toolbar);

        recyclerView = parent.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(parent.getContext(), RecyclerView.VERTICAL, false));;

        adapter = new ItemAdapter();
        recyclerView.setAdapter(adapter);
    }
}

File: app/src/main/java/com/instacart/android/challenges/MainActivity.kt
package com.instacart.android.challenges

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.viewModels
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.ui.Modifier

class MainActivity : ComponentActivity() {

    private val viewModel: MainActivityViewModel by viewModels()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        viewModel.setStateUpdateListener(
            object : MainActivityViewModel.UpdateListener {
                override fun onUpdate(state: ItemListViewState) {
                    renderItemList(state)
                }
            }
        )
    }

    @OptIn(ExperimentalMaterial3Api::class)
    private fun renderItemList(state: ItemListViewState) {
        setContent {
            Scaffold(
                topBar = {
                    TopAppBar(title = { Text(state.title) })
                }
            ) { padding ->
                Column(
                    modifier = Modifier
                        .padding(padding)
                ) {

                }
            }
        }
    }
}